src/cargo/core/resolver/mod.rs

```diff
@@ -256,12 +257,43 @@ impl<'a> Iterator for DepsNotReplaced<'a> {
     }
 }
 
+enum RcList<T> {
+    Data(Rc<(T, RcList<T>)>),
+    Empty,
+}
+
+impl<T> RcList<T> {
+    fn new() -> RcList<T> {
+        RcList::Empty
+    }
+
+    fn push(&mut self, data: T) {
+        let node = Rc::new((data, mem::replace(self, RcList::Empty)));
+        *self = RcList::Data(node);
+    }
+}
+
+// Not derived to avoid `T: Clone`
+impl<T> Clone for RcList<T> {
+    fn clone(&self) -> RcList<T> {
+        match *self {
+            RcList::Data(ref d) => RcList::Data(d.clone()),
+            RcList::Empty => RcList::Empty,
+        }
+    }
+}
+
+enum GraphNode {
+    Add(PackageId),
+    Link(PackageId, PackageId),
+}
+
 #[derive(Clone)]
 struct Context<'a> {
-    activations: HashMap<(String, SourceId), Vec<Rc<Summary>>>,
-    resolve_graph: Graph<PackageId>,
+    activations: HashMap<String, HashMap<SourceId, Vec<Summary>>>,
```

Comment:
Looks like there is almost always a single activation for a packages, so we could use a `SmallMap` here, but looks like there isn't one on crates.io? Should be pretty stratforword to roll one on top of smallvec.